src/ql/pass/map/qubits/place_mip/detail/impl.cc

```diff
@@ -0,0 +1,453 @@
+    model->lp_.sense_ = ObjSense::kMinimize;
+    model->lp_.offset_ = 0;
+
+    model->lp_.col_cost_ = std::vector<double>(model->lp_.num_col_, 0.);
+    std::fill(model->lp_.col_cost_.begin() + nfac * qubitsCount, model->lp_.col_cost_.end(), .1);
+
+    model->lp_.col_lower_ = std::vector<double>(model->lp_.num_col_, 0.);
+
+    model->lp_.col_upper_ = std::vector<double>(model->lp_.num_col_, 1.);
+    std::fill(model->lp_.col_upper_.begin() + nfac * qubitsCount, model->lp_.col_upper_.end(), 1.0e5);
+
+    model->lp_.row_lower_ = std::vector<double>(model->lp_.num_row_, 0.);
+    std::fill(model->lp_.row_lower_.begin() + qubitsCount, model->lp_.row_lower_.begin() + qubitsCount + nfac, 1.);
+    std::fill(model->lp_.row_lower_.begin() + qubitsCount + nfac, model->lp_.row_lower_.end(), -1.0e5);
+
+    model->lp_.row_upper_ = std::vector<double>(model->lp_.num_row_, 1.);
+    
+    auto costmax = computeCostMax(refcount);
+
+    for (UInt i = 0; i < nfac; ++i) {
+        for (UInt k = 0; k < qubitsCount; ++k) {
+            model->lp_.row_upper_[qubitsCount + nfac + i * qubitsCount + k] = costmax[i][k];
+        }
+    }
+
+    // model->lp_.a_matrix_ is a sparse matrix, non-zero entries are described using 3 vectors index_, value_ and start_.
+    model->lp_.a_matrix_.num_col_ = model->lp_.num_col_;
+    model->lp_.a_matrix_.num_row_ = model->lp_.num_row_;
+    model->lp_.a_matrix_.format_ = MatrixFormat::kRowwise;
+    model->lp_.a_matrix_.start_ = std::vector<int>();
+
+    for (UInt k = 0; k < qubitsCount; ++k) {
+        model->lp_.a_matrix_.start_.push_back(model->lp_.a_matrix_.value_.size());
+        model->lp_.a_matrix_.value_.insert(model->lp_.a_matrix_.value_.end(), nfac, 1);
+        for (UInt i = 0; i < nfac; ++i) {
+            model->lp_.a_matrix_.index_.push_back(i * qubitsCount + k);
+        }
+    }
+
+    for (UInt i = 0; i < nfac; ++i) {
+        model->lp_.a_matrix_.start_.push_back(model->lp_.a_matrix_.value_.size());
+        model->lp_.a_matrix_.value_.insert(model->lp_.a_matrix_.value_.end(), qubitsCount, 1);
+        for (UInt k = 0; k < qubitsCount; ++k) {
+            model->lp_.a_matrix_.index_.push_back(i * qubitsCount + k);
+        }
+    }
+
+    for (UInt i = 0; i < nfac; ++i) {
+        for (UInt k = 0; k < qubitsCount; ++k) {
+            model->lp_.a_matrix_.start_.push_back(model->lp_.a_matrix_.value_.size());
+            for (UInt j = 0; j < nfac; ++j) {
+                for (UInt l = 0; l < qubitsCount; ++l) {
+                    QL_ASSERT(distanceProvider(k, l) < utils::MAX && "All qubits in the topology should be connected");
+                    double newValue = refcount[i][j] * distanceProvider(k, l);
+                    if (j == i && l == k) {
+                        newValue += costmax[i][k];
+                    }
+
+                    // This describes a sparse matrix; 0s don't have to be added.
+                    if (std::abs(newValue) > EPSILON) {
```

Comment:
EPSILON is small, while 1.0e5 is "infinity" (but it must not be too large, otherwise HiGHS complains... I've been exchanging emails with the dev team of HiGHS, I can ask them what exactly should be used).